### PR TITLE
Caching blog post

### PR DESCRIPTION
**Pull request by mmccarty**
- [x] Add example notebooks
- [x] Incorporate initial feedback

### CODE REVIEW

File: caching.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+## Introduction
+
+Intake provides easy access data sources from remote / cloud storage, however for large files the 
+cost of downloading files everytime data is read can be extremely high. We have developed a caching 
+strategy to store and manage data sources on the local file system.
```
Comment (martindurant) [inline caching.md]: I would add the phrase "download once, read many times" somewhere in here.

File: caching.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        type: file
+    args:
+      urlpath: 's3://datashader-data/nyc_taxi.csv'
+      storage_options: {'anon': True}
+```
+
+Glob strings can be used for the ``urlpath``, which will be expanded to the individual files 
+listed in the base directory of the path. Each file will be downloaded in parallel and referrenced
+in the metadata.
+
```
Comment (martindurant) [inline caching.md]: From the users' point of view, this happens transparently, they need know nothing about the caching policy, but they will notice that read times after the initial access are much faster.

File: caching.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+See the [documentation](https://intake.readthedocs.io/en/latest/catalog.html#caching-source-files-locally)
+for more details. 
+
+## Example Notebook
+
+[![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/mmccarty/intake-blog/updates?filepath=examples%2Fcaching.ipynb)
+
+## Future Work
+
+  * To start we have implemented a caching strategy based on individual files, therefore this solution will not handle navigating directory structures to download data sources ([PR 140](https://github.com/ContinuumIO/intake/pull/140)).
```
Comment (martindurant) [inline caching.md]: The phrasing of this bullet is a bit awkward

File: caching.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+cache has been cleared explicitly using convience methods. 
+
+See the [documentation](https://intake.readthedocs.io/en/latest/catalog.html#caching-source-files-locally)
+for more details. 
+
+## Example Notebook
+
+[![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/mmccarty/intake-blog/updates?filepath=examples%2Fcaching.ipynb)
+
+## Future Work
```
Comment (martindurant) [inline caching.md]: Call it Extension, Next Steps, What's Next...,  since some of these are done?

File: caching.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+  * To start we have implemented a caching strategy based on individual files, therefore this solution will not handle navigating directory structures to download data sources ([PR 140](https://github.com/ContinuumIO/intake/pull/140)).
+  * Many remote data sources are stored as compressed files such as tar/gz/gz2/zip. We need a decompressor which will unpack after the download has completed and return the local directory ([PR 140](https://github.com/ContinuumIO/intake/pull/140)).
+  * Extensions to the Intake CLI to manage cached data ([PR 145](https://github.com/ContinuumIO/intake/pull/145)).
+  * Cached files do not have an expiration, therefore they will need to be cleared manually to get updated data. We do generate a created date when a file is downloaded and store this in the metadata. This could be used to trigger a download using a configurable expiration duration.
+  * Checksum verification need to be explored so that updated data can be downloaded regardless of expiration.
```
Comment (martindurant) [inline caching.md]: Well, some kind of unique identifier, right? To actually do a checksum, you need to download the whole thing anyway.

File: caching.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  * An API for managing and clearing cached data at the data source level, rather than accessing the cache objects directly.
```
Comment (martindurant) [inline caching.md]: and a direct user API for the metadata too.

Which reminds me (this is the wrong place): the metadata is keyed by urlpath, but the locations by hash(url+driver)? Is this inconsistency a problem?

File: caching.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+source driver, urlpath, and cache regex to avoid collision among data sources and cache 
+specifications. Files are downloaded on the first read and will persist until the 
+cache has been cleared explicitly using convience methods. 
+
+See the [documentation](https://intake.readthedocs.io/en/latest/catalog.html#caching-source-files-locally)
+for more details. 
+
+## Example Notebook
+
+[![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/mmccarty/intake-blog/updates?filepath=examples%2Fcaching.ipynb)
```
Comment (martindurant) [inline caching.md]: .. showing the typical experience of working with the cache from from an end user's (e.g., data scientist) point of view.

**Review by martindurant — COMMENTED** [review]
In the notebooks, can call `BaseCache(None, None).clear_all()` or `CacheMetadata()(urlpath)` ? Or is that an abuse of the internal API? If not, do `c = sales.cache[0]` instead of writing out in full each time; Show something about what this object is.

Explain again where that hashed directory name comes from (perhaps even call the method directly).

Where mentioning config, mention that cache and metadata/config locations can differ. I hadn't thought to set the cache dir on the data source as opposed to in in the config. Probably should show both, and explain that the latter is equivalent to the environment variable method.

File: caching.md (hunk 1 above)
Comment (mmccarty) [reply caching.md]: You sound like a tele-marketer.

Comment (martindurant) [reply caching.md]: In a good way?

Comment (mmccarty) [reply caching.md]: Haha yes!  That's the intake caching slogan.

File: caching.md (hunk 3 above)
Comment (mmccarty) [reply caching.md]: lol yes.  very awkward

File: caching.md (hunk 7 above)
Comment (mmccarty) [reply caching.md]: I agree.  I was thinking of using an EarthML example in a Binder?
https://github.com/pyviz-topics/EarthML/blob/master/examples/catalog.yml#L5
https://github.com/pyviz-topics/EarthML/blob/master/examples/Walker_Lake.ipynb

File: caching.md (hunk 6 above)
Comment (mmccarty) [reply caching.md]: Yep, we don't have a unique identifier for metadata when using multiple cache strategies.  If that's what you mean?

Comment (martindurant) [reply caching.md]: Or even different plugins looking at the same base data: you would only get one metadata entry, but multiple directories. Is that OK? Clearing the cache on one would remove only its files, but pop the whole entry, orphaning the files of the other.

File: caching.md (hunk 7 above)
Comment (mmccarty) [reply caching.md]: @jbednar I was thinking of creating a binder for the Walker Lake notebook and adding a few more cells to demonstrate caching.  Thoughts?

File: caching.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+cost of downloading files everytime data is read can be extremely high. We have developed a 
+download once, read many times caching strategy to store and manage data sources on the 
+local file system.
+
+## How it works
+
+Intake catalogs are composed of data sources, which contain attributes descibing how to 
+load data files. To enable caching, catalog authors can now specify caching parameters 
+that will be used to download data file(s), store them in a configured local directory, 
+and return a their location on local disk. For example:
```
Comment (jbednar) [inline caching.md]: return their

File: caching.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        regex: 'datashader-data'
+        type: file
+    args:
+      urlpath: 's3://datashader-data/nyc_taxi.csv'
+      storage_options: {'anon': True}
+```
+
+Glob strings can be used for the ``urlpath``, which will be expanded to the individual files 
+listed in the base directory of the path. Each file will be downloaded in parallel and referrenced
+in the metadata. From the user's point of view, this happens transparently, they need know nothing
```
Comment (jbednar) [inline caching.md]: transparently: they

File: caching.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+about the caching policy, but they will notice that read times after the initial access are 
+much faster.
+
+Downloaded files are stored locally in the cache directory, specified in the configuration 
+file, environment variable, or at runtime. The default location for the cache directory 
+is ``~/.intake/cache``. Inside this directory are subdirectories for each data source found in 
+the catalog file that has been cached. These directories are named by hashing the data 
+source driver, urlpath, and cache regex to avoid collision among data sources and cache 
+specifications. Files are downloaded on the first read and will persist until the 
+cache has been cleared explicitly using convience methods.
```
Comment (jbednar) [inline caching.md]: convenience

File: caching.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+See the [documentation](https://intake.readthedocs.io/en/latest/catalog.html#caching-source-files-locally)
+for more details. 
+
+## Example Notebook
+
+[![Binder](https://mybinder.org/badge.svg)](https://mybinder.org/v2/gh/mmccarty/intake-blog/updates?filepath=examples%2Fcaching.ipynb)
+
+## What's Next
+
+  * This caching strategy is based on individual files, therefore this solution will not handle navigating urlpaths referrencing directory structures. This is addressed in [PR 140](https://github.com/ContinuumIO/intake/pull/140).
```
Comment (jbednar) [inline caching.md]: referencing

File: caching.md [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+  * Many remote data sources are stored as compressed files such as tar/gz/gz2/zip. We need a decompressor which will unpack after the download has completed and return the local directory ([PR 140](https://github.com/ContinuumIO/intake/pull/140)).
+  * Extensions to the Intake CLI to manage cached data ([PR 145](https://github.com/ContinuumIO/intake/pull/145)).
+  * Cached files do not have an expiration, therefore they will need to be cleared manually to get updated data. We do generate a created date when a file is downloaded and store this in the metadata. This could be used to trigger a download using a configurable expiration duration.
+  * A method to determine a updates for remote data files need to be explored so that cached data can be refreshed regardless of expiration. Checksum verification is problematic since that would require downloading remote files with each check, which defeats the purpose of caching in the first place.
```
Comment (jbednar) [inline caching.md]: determine updates

Can you use modified date and size as an approximation to a checksum?  Seems like that information will sometimes be available, and could be used if it is.

File: examples/caching.ipynb [hunk 13]
```diff
@@ -0,0 +1,424 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Intake Caching\n",
+    "\n",
+    "This notebook shows a simple demonstration of how you would use and manage caching with Intake to avoid repeated downloads to large data files.\n",
```
Comment (jbednar) [inline examples/caching.ipynb]: of large

File: examples/caching.ipynb [hunk 14]
```diff
@@ -0,0 +1,424 @@
...
+   "outputs": [],
+   "source": [
+    "%time df = sales.read()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now lets read the data again.  Notice, the read is fast this time thanks to local caching."
```
Comment (jbednar) [inline examples/caching.ipynb]: let's

File: examples/caching.ipynb [hunk 15]
```diff
@@ -0,0 +1,424 @@
...
+    "cache.clear_all?"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Cache directory is configurable\n",
+    "\n",
+    "The config and cache metadata are stored in ``~/.intake``.  By default, the cache directory is located at ``~/.intake/cache``, however it can be set to a seperate location specified in the config file, an environment variable, or at runtime.  Here it is set at runtime."
```
Comment (jbednar) [inline examples/caching.ipynb]: separate

File: examples/caching.ipynb [hunk 16]
```diff
@@ -0,0 +1,424 @@
...
+    "sales.cache_dirs"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Disable Caching\n",
+    "\n",
+    "Caching can be disabled globaly in the ``intake.config``."
```
Comment (jbednar) [inline examples/caching.ipynb]: globally

File: caching.md (hunk 10 above)
Comment (jbednar) [reply caching.md]: So there's no easy way to have the cached files put into the current directory?  In most cases I'd rather have the files associated with the notebook or project where I'm working, not dumped into an ever-increasing global cache heap. I can see an example below of using get_cwd() explicitly to put it in the current directory, but that's a lot of cruft to put into every notebook.  Is it possible to have some very simple flag for configuring intake (something `intake.local_cache=True`) to store cached items in a subdirectory like `.intake_cache/` in the current directory?

**Review by jbednar — APPROVED** [review]
Looks great!  I've added notes on various typos, plus a few more substantive issues.

File: caching.md (hunk 10 above)
Comment (mmccarty) [reply caching.md]: I think this is a good idea.  I'll create an issue and we can discuss there.

Comment (mmccarty) [reply caching.md]: [Issue created](https://github.com/ContinuumIO/intake/issues/151)

File: caching.md (hunk 12 above)
Comment (mmccarty) [reply caching.md]: [Create issue](https://github.com/ContinuumIO/intake/issues/152) to track discussion

**Pull request merged**